pennylane/decomposition/decomposition_rule.py

```diff
@@ -0,0 +1,327 @@
+    if isinstance(op_type, CompressedResourceOp):
+        return op_type
+    if not issubclass(op_type, Operator):
+        raise TypeError(
+            "The keys of the dictionary returned by the resource function must be a subclass of "
+            "Operator or a CompressedResourceOp constructed with qml.resource_rep"
+        )
+    try:
+        return resource_rep(op_type)
+    except TypeError as e:
+        raise TypeError(
+            f"Operator {op_type.__name__} has non-empty resource_keys. A resource "
+            f"representation must be explicitly constructed using qml.resource_rep"
+        ) from e
+
+
+_decompositions = defaultdict(list)
+"""dict[type, list[DecompositionRule]]: A dictionary mapping operator types to decomposition rules."""
+
+
+def add_decomps(op_type: Type[Operator], *decomps: DecompositionRule) -> None:
+    """Globally registers new decomposition rules with an operator class.
+
+    .. note::
+
+        This function is only relevant when the new experimental graph-based decomposition system
+        (introduced in v0.41) is enabled via ``qml.decompositions.enable_graph()``. This new way of
+        doing decompositions is generally more performant and accommodates multiple alternative
+        decomposition rules for an operator. In this new system, custom decomposition rules are
+        defined as quantum functions, and it is currently required that every decomposition rule
+        declares its required resources using ``qml.register_resources``
+
+    In the new system of decompositions, multiple decomposition rules can be registered for the
+    same operator class. The specified decomposition rules serve as alternative decomposition rules
+    that can be chosen if they lead to a more efficient decomposition of the operator.
+
+    Args:
+        op_type: the operator type for which new decomposition rules are specified.
+        decomps (DecompositionRule): new decomposition rules to add to the given ``op_type``.
+            A decomposition is a quantum function registered with a resource estimate using
+            ``qml.register_resources``.
+
+    .. seealso:: :func:`~pennylane.register_resources` and :class:`~pennylane.list_decomps`
+
+    **Example**
+
+    .. code-block:: python
+
+        import pennylane as qml
+
+        qml.decompositions.enable_graph()
+
+        @qml.register_resources({qml.RZ: 2, qml.RX: 1, qml.GlobalPhase: 1})
+        def my_hadamard1(wires):
+            qml.RZ(np.pi / 2, wires=wires)
+            qml.RX(np.pi / 2, wires=wires)
+            qml.RZ(np.pi / 2, wires=wires)
+            qml.GlobalPhase(-np.pi / 2, wires=wires)
+
+        @qml.register_resources({qml.RZ: 1, qml.RY: 1, qml.GlobalPhase: 1})
+        def my_hadamard2(wires):
+            qml.RZ(np.pi, wires=wires)
+            qml.RY(np.pi / 2, wires=wires)
+            qml.GlobalPhase(-np.pi / 2)
+
+        qml.add_decomps(qml.Hadamard, my_hadamard1, my_hadamard2)
```

Comment:
Oh okay I thought we were going with this: 

```
qml.add_decomps(qml.H, [my_h1, mh2])
```

I really don't have a preference, I guess? Is having a variable amount of arguments better/worse technically? Any concerns with jit compatibility ?

Reply:
I thought so too, but I changed it to this API to be consistent with the PRD.

Reply:
Oh sorry 😆 I meant for this to be a list in the PRD 🤦. My only concern there is compatibility with capture/jit, but I think since it's a pytree it should be fine, right?

Reply:
I don't think `add_decomps` will ever happen in a capture/jit context?

Reply:
We're not passing any numeric data in here?

Reply:
If there are no technical concerns, I'm fine with variable number of `args` 😁